ninja-core/src/main/java/ninja/cache/CacheProvider.java

```diff
@@ -25,84 +27,51 @@
- * 
- * 
- * @author ra
- *
  */
 @Singleton
 public class CacheProvider implements Provider<Cache> {
-
-    private final NinjaProperties ninjaProperties;
-    private final Injector injector;
-
-    private final Cache cache;
-    private final Logger logger;
-
+    static private final Logger logger = LoggerFactory.getLogger(CacheProvider.class);
+    
+    private final ImplFromPropertiesFactory<Cache> factory;
+    private final Supplier<Cache> supplier;
+    
     @Inject
-    public CacheProvider(
-                         Injector injector, 
-                         NinjaProperties ninjaProperties, 
-                         Logger logger) {
-        
-        this.ninjaProperties = ninjaProperties;
-        this.injector = injector;
-        this.logger = logger;
+    public CacheProvider(Injector injector, NinjaProperties ninjaProperties) {
+        this.factory = new ImplFromPropertiesFactory<>(
+            injector,
+            ninjaProperties,
+            NinjaConstant.CACHE_IMPLEMENTATION,
+            Cache.class,
+            "ninja.cache.CacheEhCacheImpl",
+            true,
+            logger);
         
-      
-        Class<? extends Cache> cacheClass = null;
-
-        String cacheImplClassName = ninjaProperties.get(NinjaConstant.CACHE_IMPLEMENTATION);
-        
-        if (cacheImplClassName != null) {
-            try {
-
-                Class<?> clazz = Class.forName(cacheImplClassName);
-                cacheClass = clazz.asSubclass(Cache.class);
-
-                logger.info("Using the {} as implementation for caching.",  cacheClass);
-
-            } catch (ClassNotFoundException e) {
-                
-                throw new RuntimeException(
-                        "Class defined in configuration " + NinjaConstant.CACHE_IMPLEMENTATION +
-                        "not found (" + cacheClass + ")", e);
-                
-            } catch (ClassCastException e) {
-                
-                throw new RuntimeException(
-                        "Class defined in configuration " 
-                                + NinjaConstant.CACHE_IMPLEMENTATION +
-                                "is not an instance of interface cache (" 
-                                + cacheClass + ")", e);
+        // lazy singleton
+        this.supplier = Suppliers.memoize(new Supplier<Cache>() {
+            @Override
+            public Cache get() {
+                return factory.create();
             }
-        }
-
-        if (cacheClass == null) {
-                // load default implementation
-                cacheClass = CacheEhCacheImpl.class;
-                logger.info("Using default eh cache implementation. ({}) ", cacheClass);
-            
-        }
-
-        cache = injector.getInstance(cacheClass);
-      
+        });
     }
 
     @Override
     public Cache get() {
-        // only called once => reference cached by guice...
-        return cache;
-        
+        return this.supplier.get();
```

Comment:
Do we really need a supplier here? I actually thought Guice is managing that for us...